waiter/src/waiter/core.clj

```diff
@@ -1249,11 +1249,20 @@
 
    :wrap-service-discovery-fn (pc/fnk [discover-service-parameters-fn]
                                 (fn wrap-service-discovery-fn
-                                  [handler]
+                                  [handler & {:keys [require-bypass-token-host] :or {require-bypass-token-host false}}]
                                   (fn [{:keys [headers] :as request}]
                                     ;; TODO optimization opportunity to avoid this re-computation later in the chain
-                                    (let [discovered-parameters (discover-service-parameters-fn headers)]
-                                      (handler (assoc request :waiter-discovery discovered-parameters))))))})
+                                    (let [{:strs [host]} headers
+                                          hostname (-> host (str/split #":") first)
+                                          {:keys [service-description-template token] :as discovered-parameters} (discover-service-parameters-fn headers)]
+                                      (handler (cond-> request
+                                                 ; require-bypass-token-host flag only does waiter discovery when the hostname is the token
+                                                 ; and the token is in bypass mode. Usually this flag is enabled so that discovery is done before
+                                                 ; the wrap-secure-request-fn middleware because bypass tokens may authenticate differently
+                                                 (or (not require-bypass-token-host)
+                                                     (and (= token hostname)
+                                                          (= "true" (get-in service-description-template ["metadata" "waiter-proxy-bypass-opt-in"]))))
```

Comment:
Under what circumstances does discovery need to be conditional on `require-bypass-token-host`? 
Feels like the auth logic need to inspect host headers and determine which auth mechanism to use.
In other words, why can't we do discovery and rely on using the discovered parameters as needed when performing auth?